pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -316,6 +316,242 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
+			if !podutil.IsRestartableInitContainer(&container) {
+				continue
+			}
+			if container.Resources.Requests.Cpu().IsZero() {
+				hasSharedAppContainers = true
+				break
+			}
+		}
+	}
+
+	if sumOfLongRunningExclusiveCPUs >= podTotalCPUs && hasSharedAppContainers {
+		return fmt.Errorf("pod rejected, sum of exclusive container cpu requests equals pod budget, leaving no cpus for shared containers")
+	}
+
+	// Check for empty shared pool for standard init containers.
+	sumOfSidecarExclusiveCPUs := 0
+	for _, container := range pod.Spec.InitContainers {
+		if podutil.IsRestartableInitContainer(&container) {
+			sumOfSidecarExclusiveCPUs += p.guaranteedCPUs(logger, pod, &container)
+		}
+	}
+
+	poolForStandardInits := podTotalCPUs - sumOfSidecarExclusiveCPUs
+	maxStandardInitExclusiveCPUs := 0
+	hasSharedStandardInitContainer := false
+	for _, container := range pod.Spec.InitContainers {
+		if !podutil.IsRestartableInitContainer(&container) {
+			cpus := p.guaranteedCPUs(logger, pod, &container)
+			if cpus > 0 {
+				if cpus > maxStandardInitExclusiveCPUs {
+					maxStandardInitExclusiveCPUs = cpus
+				}
+			} else {
+				hasSharedStandardInitContainer = true
+			}
+		}
+	}
+
+	// This check ensures that the pool of CPUs available to standard init containers
+	// (the total pod budget minus any sidecars) is large enough to satisfy the
+	// request of the largest standard init container that requires exclusive CPUs.
+	// Since standard init containers run sequentially, we only need to fit the largest one.
+	if maxStandardInitExclusiveCPUs > poolForStandardInits {
+		return fmt.Errorf("pod rejected, largest exclusive init container requests %d cpus, but only %d available", maxStandardInitExclusiveCPUs, poolForStandardInits)
+	}
+
+	// This check ensures that if there are any standard init containers that need a
+	// shared pool, that pool is not empty. An empty pool would occur if the pod
+	// budget is fully consumed by sidecars.
+	if hasSharedStandardInitContainer && poolForStandardInits == 0 {
+		return fmt.Errorf("pod rejected, pod has shared init containers but no cpus available for them")
+	}
+
+	return nil
+}
+
+// This function is the entry point for pod-level resource allocation.
+// It's called once per pod by the Topology Manager's pod-scope admit handler.
+// The logic here allocates a single "bubble" of CPUs for the entire pod
+// and then partitions that bubble among the containers.
+func (p *staticPolicy) AllocatePod(logger logr.Logger, s state.State, pod *v1.Pod, hint topologymanager.TopologyHint) error {
+	// Centralized check: Do nothing if this pod is not managed at the pod level.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResourceManagers) || !resourcehelper.IsPodLevelResourcesSet(pod) {
+		return nil
+	}
+
+	// 1. Validate for the "empty shared pool" case.
+	if err := p.validatePodScopeResources(logger, pod); err != nil {
+		return &admission.EmptyPodSharedPoolError{Msg: err.Error()}
```

Comment:
If we create this error just and only from an error, it makes sense to add a builder function which takes an error as only argument and sets the internal `Message` accordingly; at that point we can also have the `message` member private and this is an extra nicety.

Reply:
Done.